File: modules/server/connect.lua
AddEventHandler("playerConnecting", function (playerName, setKickReason, deferrals)
    source = tonumber(source)
    local ipv4 = Util.GetIdentifier(source, "ip:"):sub(4)
    if ServerConfig.Modules.Connect.NameFilter.enabled or ServerConfig.Modules.Connect.NoVPN.enabled then
        deferrals.defer()
        Citizen.Wait(0)
    end

    if not Util.IsAlphaNumeric(tostring(playerName), " .,?!+-*:/~#_<>|^") and ServerConfig.Modules.Connect.NameFilter.enabled then
        deferrals.done(ServerConfig.Modules.Connect.NameFilter.rejectionMsg)
    end

    if ServerConfig.Modules.Connect.NoVPN.enabled then
        PerformHttpRequest("https://blackbox.ipinfo.app/lookup/" .. ipv4, function(errorCode, usingVPN, resultHeaders)
            if not usingVPN == "N" then
                deferrals.done()
            else
                deferrals.done(ServerConfig.Modules.Connect.NoVPN.rejectionMsg)
            end
        end)
    end
end)

File: modules/server/connect.js
onNet("playerConnecting", async (playerName, _, deferrals) => {
    const ipv4 = GetPlayerIdentifierByType(source, "ip").slice(4);
    if (serverConfig.Modules.Connect.NameFilter.enabled || serverConfig.Modules.Connect.NoVPN.enabled) {
        deferrals.defer();
        await Delay(0);
    }

    if (serverConfig.Modules.Connect.NameFilter.enabled && !(/^[a-zA-Z0-9!@#\$%\^\&*\)\(+=._-]+$/g.test(playerName))) {
        deferrals.done(ServerConfig.Modules.Connect.NameFilter.rejectionMsg);
    }

    if (serverConfig.Modules.Connect.NoVPN.enabled) {
        try {
            const request = await axios.get(`https://blackbox.ipinfo.app/lookup/${ipv4}`, config);
            if (request.data[0] == "N") {
                deferrals.done();
            } else {
                deferrals.done(serverConfig.Modules.Connect.NoVPN.rejectionMsg);
            }
        } catch (err) {
            console.log(err);
        }
    } else {
        deferrals.done();
    }
});
